### PR TITLE
Adding dynamic-client interface

### PR DESCRIPTION
**Pull request by barney-s**
Add support for dynamic client interface.
Can be extended to support typed client if needed later.

Avoiding taking a dependency on controller-runtime which has similar client wrapping over dynamic interface.

### CODE REVIEW

File: internal/pkg/client/client.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	i, err := r.List(options)
+	if err != nil {
+		return err
+	}
+	u.Items = i.Items
+	u.Object = i.Object
+	return nil
+}
+
+func (uc *client) UpdateStatus(_ context.Context, obj runtime.Object) error {
```
Comment (pwittrock) [inline internal/pkg/client/client.go]: nit: docs

File: internal/pkg/client/client.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+// client is a client.Client that reads and writes directly from/to an API server.  It lazily initializes
+// new clients at the time they are used, and caches the client.
+type client struct {
+	client     dynamic.Interface
+	restMapper meta.RESTMapper
+}
+
+// Create implements client.Client
+func (uc *client) Create(_ context.Context, obj runtime.Object, options metav1.CreateOptions) error {
+	u, ok := obj.(*unstructured.Unstructured)
```
Comment (pwittrock) [inline internal/pkg/client/client.go]: This code is copied for each function.  WDYT of pulling it into its own function?

File: internal/pkg/client/client.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	r, err := uc.getResourceInterface(u.GroupVersionKind(), u.GetNamespace())
+	if err != nil {
+		return err
+	}
+	err = r.Delete(u.GetName(), options)
+	return err
+}
+
+// Patch implements client.Client
+func (uc *client) Patch(_ context.Context, obj runtime.Object, patch Patch, options metav1.PatchOptions) error {
```
Comment (pwittrock) [inline internal/pkg/client/client.go]: WYDT of adding an `Apply` function, which wraps Patch?

**Comment by pwittrock** [comment]
Look good.  Plz add tests.  You can use the test framework to make integration tests.

File: internal/pkg/client/client.go (hunk 2 above)
Comment (barney-s) [reply internal/pkg/client/client.go]: done.

**Comment by barney-s** [comment]
@Liujingfang1 PTAL

File: internal/pkg/client/client.go (hunk 3 above)
Comment (barney-s) [reply internal/pkg/client/client.go]: Added an Apply() please review.

File: internal/pkg/client/client.go (hunk 1 above)
Comment (barney-s) [reply internal/pkg/client/client.go]: Done.

**Comment by Liujingfang1** [comment]
@barney-s Looks good to me.

**Comment by Liujingfang1** [comment]
@pwittrock Can we merge this PR? Then I can use the dynamic client from it.

**Pull request merged**